Sources/BowEffects/Data/Schedule.swift

```diff
@@ -0,0 +1,768 @@
+            }
+        )
+    }
+    
+    /// Chooses between two schedules with different inputs and the same output.
+    ///
+    /// - Parameter that: Schedule to be composed with this one.
+    public func choose<State2, C>(_ that: Schedule<R, C, State2, B>) -> Schedule<R, Either<A, C>, (State, State2), B> {
+        self.choose(that).map { b in b.fold(id, id) }
+    }
+    
+    /// Composes this schedule with another one, providing a new schedule that continues as long as both continue, using the maximum delay of the two schedules.
+    ///
+    /// - Parameter that: Schedule to be composed with this one.
+    public func zip<State2, C>(_ that: Schedule<R, A, State2, C>) -> Schedule<R, A, (State, State2), (B, C)> {
+        self.and(that)
+    }
+    
+    /// Composes this schedule with another one, providing a new schedule that continues as long as both continue, using the maximum delay of the two schedules, and ignoring the output of this schedule.
+    ///
+    /// - Parameter that: Schedule to be composed with this one.
+    public func zipRight<State2, C>(_ that: Schedule<R, A, State2, C>) -> Schedule<R, A, (State, State2), C> {
+        self.and(that).map { x in x.1 }
+    }
+    
+    /// Composes this schedule with another one, providing a new schedule that continues as long as both continue, using the maximum delay of the two schedules, and ignoring the output of the provided schedule.
+    ///
+    /// - Parameter that: Schedule to be composed with this one.
+    public func zipLeft<State2, C>(_ that: Schedule<R, A, State2, C>) -> Schedule<R, A, (State, State2), B> {
+        self.and(that).map { x in x.0 }
+    }
+    
+    /// Pipes the output of this schedule as the input of the provided one. Effects described by this schedule will always be performed before the effects described by the second schedule.
+    ///
+    /// - Parameter that: Schedule to be piped after this one.
+    public func pipe<State2, C>(_ that: Schedule<R, B, State2, C>) -> Schedule<R, A, (State, State2), C> {
+        Schedule<R, A, (State, State2), C>(
+            initial: self.initial.zip(that.initial),
+            extract: { a, s in that.extract(self.extract(a, s.0), s.1) },
+            update: { a, s in
+                let s1 = EnvIO<R, Unit, State>.var()
+                let s2 = EnvIO<R, Unit, State2>.var()
+                
+                return binding(
+                    s1 <- self.update(a, s.0),
+                    s2 <- that.update(self.extract(a, s.0), s.1),
+                    yield: (s1.get, s2.get))^
+            }
+        )
+    }
+    
+    /// Pipes the output of the provided schedule as the input of this schedule. Effects described by the provided schedule will always be performed before the effects described by this schedule.
+    ///
+    /// - Parameter that: Schedule to be piped before this one.
+    public func reversePipe<State2, C>(_ that: Schedule<R, C, State2, A>) -> Schedule<R, C, (State2, State), B> {
```

Comment:
Maybe it is a dummy question, but I reverse like:
`public func reversePipe<State2, C>(_ that: Schedule<R, C, State2, B>) -> Schedule<R, C, (State2, State), A>`

Reply:
No, this is like the operators `<<<` and `>>>`, with the following signatures:

`>>>`: `(f: (A) -> B, g: (B) -> C) -> (A) -> C`
`>>>`: `(g: (B) -> C, f: (A) -> B) -> (A) -> C`

The only thing you are reversing is the order of the parameters, not their type arguments.

Reply:
Ah! Now I see the point, thanks